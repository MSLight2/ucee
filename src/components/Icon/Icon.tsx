import React from 'react';
import classnames from 'classnames';
import { getIsImage, transformSize } from '../../utils';
import './index.scss'

export interface IconProps {
  name: string,
  dot?: boolean,
  info?: string | number,
  color?: string,
  size?: string | number,
  className?: string,
  classPrefix?: string,
  containerClassName?: string
  onClick?: Function
}

const basePrefix: string = 'ucee-icon';

const Icon: React.FC<IconProps> = (props) => {
  const {
    name,
    dot,
    info,
    color,
    size,
    className,
    classPrefix = basePrefix,
    containerClassName = `${basePrefix}__box`,
    onClick
  } = props;

  const containerClass = classnames(`${containerClassName}`);

  let isImage = getIsImage(name);
  const classNames = classnames(
    `${classPrefix}`,
    {
      [`${classPrefix}-${name}`]: !isImage
    },
    className
  );
  const iconProps = {
    className: classNames,
    style: {}
  };
  if (color) {
    Object.assign(iconProps, {
      style: {
        ...iconProps.style,
        color
      }
    });
  }
  if (size) {
    Object.assign(iconProps, {
      style: {
        ...iconProps.style,
        fontSize: transformSize(size)
      }
    });
  }
  if (onClick) {
    Object.assign(iconProps, { onClick });
  }

  const IconElem = () => {
    return (
      <>
        {
          isImage ? (
            <div className={classPrefix}>
              <img src={name} className={`${classPrefix}__img`} alt="" />
            </div>
          ) : (
            <i {...iconProps}></i>
          )
        }
      </>
    )
  }

  return (
    <>
      {
        (dot || info)
          ? <div className={containerClass}>
              {dot && !info && <span className={`${classPrefix}--dot`}></span>}
              {info && Number(info) > 0 &&
                <span
                  className={`${classPrefix}--badge`}
                >
                  {Number(info) > 99 ? '99+' : info}
                </span>
              }
              <IconElem />
            </div>
          : <IconElem />
      }
    </>
  );
}

export default Icon;