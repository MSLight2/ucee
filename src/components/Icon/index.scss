@import '../../styles/var';
@import './icon.scss';

$containerClassName: 'ucee-icon__box' !default;
$basePrefix: 'ucee-icon' !default;

.#{$containerClassName}{
  position: relative;
  display: inline-block;
  font-size: $fone-size-base;
  .#{$basePrefix}--dot{
    position: absolute;
    top: 0;
    right: 0;
    width: $icon-dot-size;
    height: $icon-dot-size;
    background: $red;
    border-radius: 50%;
    z-index: 1;
    transform: translate(50%, -50%);
  }
  .#{$basePrefix}--badge{
    position: absolute;
    top: 0;
    right: 0;
    min-width: 16px;
    padding: 0 3px;
    background: $red;
    border-radius: 16px;
    line-height: 1.2;
    color: $white;
    box-sizing: border-box;
    transform: translate(50%, -50%);
  }
}
.#{$basePrefix}__img{
  object-fit: contain;
  width: 2.5em;
  height: 2.5em;
}