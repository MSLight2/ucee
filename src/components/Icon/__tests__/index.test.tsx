import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Icon from '../index';

describe('Icon', () => {
  test('should render icon with color prop', () => {
    const { container } = render(<Icon name="my" color="#f90" />);
    expect(container).toMatchSnapshot();
  });

  test('should render icon with size prop', () => {
    const { container } = render(<Icon name="my" size={30} />);
    expect(container).toMatchSnapshot();
  });

  test('should render icon with dot prop', () => {
    const { container } = render(<Icon name="my" dot />);
    expect(container.querySelector('.ucee-icon--dot')).toBeTruthy();
  });

  test('should render icon with info prop', () => {
    const { container, rerender } = render(<Icon name="my" info={20} />);
    expect(container.querySelector('.ucee-icon--badge')?.innerHTML).toEqual('20');
    rerender(<Icon name="my" info={200} />);
    expect(container.querySelector('.ucee-icon--badge')?.innerHTML).toEqual('99+');
  });

  test('should trigger onClick', () => {
    const onClick = jest.fn();
    const { container } = render(<Icon name="my" onClick={onClick} />);

    fireEvent.click(container.querySelector('.ucee-icon') as Element);
    expect(onClick).toHaveBeenCalled();
  });

  test('should render url image', () => {
    const { container } = render(<Icon name="https://img.yzcdn.cn/vant/user-active.png" />);
    expect(container.getElementsByTagName('img').length).toEqual(1);
  });
});