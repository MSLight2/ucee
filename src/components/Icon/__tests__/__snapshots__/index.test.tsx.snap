// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render icon with color prop 1`] = `
<div>
  <i
    class="ucee-icon ucee-icon-my"
    style="color: rgb(255, 153, 0);"
  />
</div>
`;

exports[`Icon should render icon with size prop 1`] = `
<div>
  <i
    class="ucee-icon ucee-icon-my"
    style="font-size: 30px;"
  />
</div>
`;
